src/main/java/fi/aalto/cs/intellij/services/CourseInformation.java

```diff
@@ -0,0 +1,106 @@
+package fi.aalto.cs.intellij.services;
+
+import com.intellij.openapi.components.ServiceManager;
+
+import java.io.FileNotFoundException;
+import java.io.FileReader;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+import org.json.JSONException;
+import org.json.JSONObject;
+import org.json.JSONTokener;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class CourseInformation {
```

Comment:
Yes, your understanding is spot on. The main issue I see with this approach is thread safety, which becomes a relevant issue once we allow the user to load another course.

I agree that the current approach doesn't quite "feel right". Your approach sounds good, although it raises the question of where the `currentlyOpenCourse` instance is stored, to which I suspect that the answer is some bigger plugin level singleton (`PluginSettings` or something similar), which sort of just defers the problem. But taking a more OO approach for the `CourseInformation` class and making it a "regular" class sounds good to me. Let's discuss this tomorrow before I refactor it.

Coupling the class with the IntelliJ service manager guarantees that only a single instance of the class exists (which is initialized on-demand in a thread safe manner) for the entire IntelliJ application, which is convenient for this class. However, you're correct in that all of this can be done with "plain old Java".